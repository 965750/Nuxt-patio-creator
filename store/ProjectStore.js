import { defineStore } from 'pinia'
import { useInterfaceStore } from './InterfaceStore'
import { useNotification } from '@kyvg/vue3-notification'
import { projectFirestore, timestamp } from '@/firebase/config'
import { collection, getDocs, getDoc, updateDoc, addDoc, doc, deleteDoc } from 'firebase/firestore'

export const useProjectStore = defineStore('ProjectStore', {
  state: () => ({
    projects: [],
    door: {
      height: 250,
      width: 120,
      doors: 2,
      color: '#712f00',
      name: ''
    },
  }),
  actions: {
    updateDoor ({attr, value}) {
        this.door[attr] = value
    },
    async saveProjectUpdate (localName) {
        const interfaceStore = useInterfaceStore()
        const { notify } = useNotification()
        const router = useRouter()

        try {
            const docRef = doc(projectFirestore, 'savedProjects', this.door.id)
                
            await updateDoc(docRef, { 
                ...this.door,
                name: localName,
                updatedAt: timestamp
            })

            this.setDefaultDoor()

            interfaceStore.switchIsLoading(false)
            setTimeout(() => {
                notify({
                    text: 'Project updated',
                    type: 'success'
                })
            }, 500)

            router.push('/')
        } catch(error) {
            notify({
                text: "Error, try again later",
                type: "error"
            })
            console.log(error)
            interfaceStore.switchIsLoading(false)
        }
    },
    async saveProject (localName) {
        if (localName === this.door.name) {
            this.saveProjectUpdate(localName)
        } else {
            const interfaceStore = useInterfaceStore()
            const { notify } = useNotification()
            const router = useRouter()

            try {
                const projectCol = collection(projectFirestore, 'savedProjects')
                await addDoc(projectCol, { 
                    ...this.door,
                    name: localName,
                    updatedAt: timestamp
                })

                this.setDefaultDoor()
                interfaceStore.switchIsLoading(false)
                setTimeout(() => {
                    notify({
                        text: 'Project saved',
                        type: 'success'
                    })
                }, 500)

                router.push('/')
            } catch(error) {
                notify({
                    text: "Error, try again later",
                    type: "error"
                })
                console.log(error)
                interfaceStore.switchIsLoading(false)
            }
        }
    },
    async getProjects () {
        const interfaceStore = useInterfaceStore()
        interfaceStore.switchIsLoading(true)
        const { notify } = useNotification()
    
        try {
            const projectCol = collection(projectFirestore, 'savedProjects')
            const projectsDocs = await getDocs(projectCol)
            
            this.projects = projectsDocs.docs.map(doc => {
                return {...doc.data(), id: doc.id }
            })
            interfaceStore.switchIsLoading(false)
        } catch(error) {
            notify({
                text: "Error, try again later",
                type: "error"
            })
            console.log(error)
            interfaceStore.switchIsLoading(false)
        }
    },
    async deleteProject (id) {
        const { notify } = useNotification()

        try {
            const docRef = doc(projectFirestore, 'savedProjects', id)
                
            await deleteDoc(docRef)
            this.getProjects()

            notify({
                text: "Project deleted",
                type: "success"
            })
        } catch(error) {
            notify({
                text: "Error, try again later",
                type: "error"
            })
            console.log(error)
        }
    },
    async loadProject (id) {
        const interfaceStore = useInterfaceStore()
        const { notify } = useNotification()
        const router = useRouter()
        
        try {
            const docRef = doc(projectFirestore, 'savedProjects', id)
            const projectDoc = await getDoc(docRef)

            if (!projectDoc.data()) {
                throw Error('wrong project id')
            }

            this.setDoor({ ...projectDoc.data(), id })
            interfaceStore.switchIsLoading(false)
            notify({
                text: "Project loaded",
                type: "success"
            })
        } catch(error) {
            setTimeout(() => {
                notify({
                    text: "Try again later, or load different project",
                    type: "error"
                })
            }, 500)
            this.setDefaultDoor()
            interfaceStore.switchIsLoading(false)
            router.push('/')
        }
    },
    setDefaultDoor () {
        this.door = {
          height: 250,
          width: 120,
          doors: 2,
          color: '#712f00',
          name: ''
        }
    },
    setDoor (payload) {
        this.door = payload
    }
  },
  getters: {
    currentDoor: (state) => state.door,
    savedProjects: (state) => state.projects,
  },
})