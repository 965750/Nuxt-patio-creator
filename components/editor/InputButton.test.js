import { describe, test, beforeEach, vitest, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import { createTestingPinia } from '@pinia/testing'
import InputButton from './InputButton.vue'
import { useProjectStore } from '@/store/ProjectStore.js'

describe('InputButton', () => {
let wrapper = null

    beforeEach(() => {
        wrapper = mount(InputButton, {
            shallow: true,
            global: {
                plugins: [
                    createTestingPinia({
                        createSpy: vitest.fn
                    })
                ]
            }
        })
    })

    test(`if user won't save too short name`, async () => {
        const input = wrapper.find('input')
        const submit = wrapper.find('button')

        await input.setValue('123')
        await submit.trigger('click')

        expect(wrapper.text()).toContain('name is too short')
    })

    test(`if user won't save empty name`, async () => {
        const input = wrapper.find('input')
        const submit = wrapper.find('button')

        await input.setValue('')
        await submit.trigger('click')

        expect(wrapper.text()).toContain('name is too short')
    })

    test(`if user won't save too long name`, async () => {
        const input = wrapper.find('input')
        const submit = wrapper.find('button')

        await input.setValue('1234567890 12345678')
        await submit.trigger('click')
        
        expect(wrapper.text()).toContain('name is too long')
    })

    test(`if user won't save symbols as project name`, async () => {
        const input = wrapper.find('input')
        const submit = wrapper.find('button')

        await input.setValue('123 %asd')
        await submit.trigger('click')
        
        expect(wrapper.text()).toContain('only letters and numbers')
    })

    test('snapshot matches', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })
})