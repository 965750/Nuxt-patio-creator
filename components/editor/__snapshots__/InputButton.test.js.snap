// Vitest Snapshot v1

exports[`InputButton > snapshot matches 1`] = `
"<div data-v-cf9c6063=\\"\\">
  <div class=\\"wrapper flex\\" data-v-cf9c6063=\\"\\"><input data-cy-name-input=\\"\\" attr=\\"\\" class=\\"wrapper__input p-2\\" type=\\"text\\" placeholder=\\"project name\\" data-v-cf9c6063=\\"\\"><button data-cy-save-project-btn=\\"\\" class=\\"wrapper__button font-bold\\" data-v-cf9c6063=\\"\\"> Save </button></div>
  <p class=\\"inputError\\" data-v-cf9c6063=\\"\\"></p>
</div>"
`;
